File: ParteC#/ProyectoProgramacion3/ProyectoProgramacion3Persistencia/DAO/reservas/DetalleReservaDaoImpl.cs
using System.Data;
using System.Data.Common;
using ProyectoProgramacion3Model.Model.reservas;

namespace ProyectoProgramacion3Persistencia.DAO.reservas;

public class DetalleReservaDaoImpl : DefaultBaseDao<DetalleReserva>, IDetalleReservaDao
{
    protected override DbCommand ComandoCrear(DbConnection conn, DetalleReserva modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_InsertDetalle_Reserva");
        SetCamposDetalle(cmd, modelo);
        CrearParametroSalida(cmd, "_id_generado", DbType.Int32);
        return cmd;
    }

    protected override DbCommand ComandoActualizar(DbConnection conn, DetalleReserva modelo)
    {
        var cmd = CrearStoredProcedure(conn, "sp_UpdateDetalle_Reserva");
        CrearParametro(cmd, "p_id_detalle_reserva", modelo.idDetalleReserva);
        SetCamposDetalle(cmd, modelo);
        return cmd;
    }

    protected override DbCommand ComandoEliminar(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_DeleteDetalle_Reserva");
        CrearParametro(cmd, "p_id_detalle_reserva", id);
        return cmd;
    }

    protected override DbCommand ComandoLeer(DbConnection conn, int id)
    {
        var cmd = CrearStoredProcedure(conn, "sp_ListDetalleReservaById");
        CrearParametro(cmd, "p_id_detalle_reserva", id);
        return cmd;
    }

    protected override DbCommand ComandoLeerTodos(DbConnection conn)
    {
        return CrearStoredProcedure(conn, "spListDetalleReserva");
    }

    protected override DetalleReserva MapearModelo(DbDataReader reader)
    {
        return new DetalleReserva
        {
            idDetalleReserva = LeerEntero(reader, "id_detalle_reserva"),
            idReserva = LeerEntero(reader, "id_reserva"),
            idServicio = LeerEntero(reader, "id_servicio"),
            cantidad = LeerEntero(reader, "cantidad"),
            subtotal = LeerDouble(reader, "subtotal")
        };
    }

    private void SetCamposDetalle(DbCommand cmd, DetalleReserva modelo)
    {
        CrearParametro(cmd, "p_id_reserva", modelo.idReserva);
        CrearParametro(cmd, "p_id_servicio", modelo.idServicio);
        CrearParametro(cmd, "p_cantidad", modelo.cantidad);
        CrearParametro(cmd, "p_subtotal", modelo.subtotal);
    }
}

File: ParteJAVA/ProyectoPro3Persistencia/src/main/java/pe/edu/pucp/proyectopro3/dao/reservas/DetalleReservaDAOImpl.java
package pe.edu.pucp.proyectopro3.dao.reservas;

import pe.edu.pucp.proyectopro3.dao.DefaultBaseDAO;
import pe.edu.pucp.proyectopro3.modelo.reservas.DetalleReserva;

import java.sql.*;

public class DetalleReservaDAOImpl extends DefaultBaseDAO<DetalleReserva> implements DetalleReservaDAO {

    @Override
    protected PreparedStatement comandoCrear(Connection conn, DetalleReserva modelo) throws SQLException {
        String sql = "{call sp_InsertDetalle_Reserva(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        setCamposDetalleReserva(cmd, modelo, 1);
        cmd.registerOutParameter(5, Types.INTEGER);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, DetalleReserva modelo) throws SQLException {
        String sql = "{call sp_UpdateDetalle_Reserva(?, ?, ?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);

        cmd.setInt(1, modelo.getIdDetalle());
        setCamposDetalleReserva(cmd, modelo, 2);

        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_DeleteDetalle_Reserva(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) throws SQLException {
        String sql = "{call sp_ListDetalleReservaById(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt(1, id);
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) throws SQLException {
        String sql = "{call spListDetalleReserva()}";
        return conn.prepareCall(sql);
    }

    @Override
    protected Integer extraerIdDesdeCallable(CallableStatement cmd) throws SQLException {
        return cmd.getInt(5);
    }

    @Override
    protected Integer extraerIdDesdeGeneratedKeys(ResultSet rs) throws SQLException {
        return rs.getInt(1);
    }

    @Override
    protected DetalleReserva mapearModelo(ResultSet rs) throws SQLException {
        DetalleReserva detalle = new DetalleReserva();

        detalle.setIdDetalle(rs.getInt("id_detalle_reserva"));
        detalle.setIdReserva(rs.getInt("id_reserva"));
        detalle.setIdServicio(rs.getInt("id_servicio"));
        detalle.setCantidad(rs.getInt("cantidad"));
        detalle.setSubtotal(rs.getDouble("subtotal"));

        return detalle;
    }

    private void setCamposDetalleReserva(CallableStatement cmd, DetalleReserva modelo, int inicio) throws SQLException {
        cmd.setInt(inicio, modelo.getIdReserva());
        cmd.setInt(inicio + 1, modelo.getIdServicio());
        cmd.setInt(inicio + 2, modelo.getCantidad());
        cmd.setDouble(inicio + 3, modelo.getSubtotal());
    }
}
